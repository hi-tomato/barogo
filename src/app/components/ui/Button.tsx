"use client";

interface ButtonProps {
  text: string;
  size?: "sm" | "md" | "lg";
  disabled?: boolean;
  type?: "button" | "submit" | "reset";
  onClick?: () => void;
  icon?: React.ReactNode;
  className?: string;
}

export default function Button({
  text,
  type,
  disabled,
  onClick,
  icon,
  className,
}: ButtonProps) {
  return (
    <button
      type={type}
      onClick={onClick}
      disabled={disabled}
      className={className}
    >
      {icon && icon}
      {text}
    </button>
  );
}
