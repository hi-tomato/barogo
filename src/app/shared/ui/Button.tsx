'use client';
import React from 'react';
import { HiArrowLeft, HiArrowRight } from 'react-icons/hi';
import { cn } from '../lib/cn';
import Loading from './Loading';

type Variant =
  | 'primary'
  | 'secondary'
  | 'outline'
  | 'gradient'
  | 'ghost'
  | 'text'
  | 'tabButton'
  | 'kakao'
  | 'popularButton'
  | 'nearbyButton';
type Size = 'sm' | 'md' | 'lg' | 'icon' | 'pill';
type Icon = 'left' | 'right' | 'back' | 'forward' | React.ReactNode;
type IconPosition = 'left' | 'center' | 'right';
type ButtonType = 'button' | 'submit' | 'reset';
interface ButtonProps {
  children?: React.ReactNode;
  onClick?: (() => void) | ((e: React.FormEvent) => void);
  disabled?: boolean;
  loading?: boolean;
  variant?: Variant;
  size?: Size;
  icon?: Icon;
  iconPosition?: IconPosition;
  className?: string;
  type?: ButtonType;
  text?: string;
  fullWidth?: boolean;
}

const buttonVariants = {
  primary: 'bg-[#1C4E80] hover:bg-[#154066] text-white transition-colors',
  secondary: 'bg-gray-500 hover:bg-gray-600 text-white transition-colors',
  outline:
    'border border-[#1C4E80]/15 text-[#1C4E80] hover:bg-[#1C4E80]/5 hover:border-[#1C4E80]/30 transition-all',
  gradient:
    'bg-gradient-to-r from-blue-500 to-blue-600 hover:shadow-md transition-all',
  ghost: 'text-gray-600 hover:bg-gray-100 transition-colors',
  text: 'text-[#1C4E80] text-center transition-colors bg-transparent border-none',
  tabButton: 'bg-[#1C4E80] text-white transition-colors',
  kakao: 'bg-[#FEE500] hover:bg-[#FDD835] text-black transition-colors',
  popularButton: 'bg-[#1C4E80] hover:bg-[#154066] text-white transition-colors',
  nearbyButton:
    'bg-[#1C4E80] hover:bg-[#154066] text-white rounded-full transition-colors',
};

const buttonSizes = {
  sm: 'px-3 py-2 text-sm rounded',
  md: 'px-4 py-2 text-base rounded-lg',
  lg: 'px-6 py-3 text-lg rounded-lg',
  icon: 'p-2 rounded-lg',
  pill: 'px-4 py-2 text-sm rounded-full',
};

const iconComponents = {
  left: null,
  right: null,
  back: HiArrowLeft,
  forward: HiArrowRight,
};

export default function Button({
  children,
  onClick,
  disabled = false,
  loading = false,
  variant = 'primary',
  size = 'md',
  fullWidth = false,
  icon,
  iconPosition = 'left',
  className = '',
  type = 'button',
  text,
}: ButtonProps) {
  const baseClasses =
    variant === 'text'
      ? 'font-semibold transition-all duration-200 disabled:opacity-50 disabled:cursor-not-allowed'
      : 'font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed';
  const variantClasses = buttonVariants[variant];
  const sizeClasses = buttonSizes[size];
  const combinedClasses = cn(
    baseClasses,
    variantClasses,
    sizeClasses,
    {
      'w-full': fullWidth,
    },
    className
  );

  const renderIcon = () => {
    if (loading) {
      return <Loading />;
    }

    if (!icon) return null;

    if (typeof icon !== 'string') {
      return icon;
    }

    const IconComponent = iconComponents[icon as keyof typeof iconComponents];

    if (icon === 'back' && IconComponent) {
      return <IconComponent size={20} />;
    }

    if (icon === 'forward' && IconComponent) {
      return <IconComponent size={20} />;
    }

    return null;
  };

  const iconElement = renderIcon();

  return (
    <button
      type={type}
      onClick={onClick}
      disabled={disabled || loading}
      className={combinedClasses}
    >
      {variant === 'text' ? (
        <>
          {iconPosition === 'left' && (
            <>
              {iconElement}
              <span className="mr-1" />
            </>
          )}
          {text || children}
          {iconPosition === 'right' && (
            <>
              <span className="ml-1" />
              {iconElement}
            </>
          )}
        </>
      ) : (
        <>
          {iconPosition === 'left' && iconElement}
          {iconPosition === 'center' ? (
            iconElement
          ) : (
            <span>{text || children}</span>
          )}
          {iconPosition === 'right' && iconElement}
        </>
      )}
    </button>
  );
}
