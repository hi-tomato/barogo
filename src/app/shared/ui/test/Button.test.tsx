import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button from '../Button';

describe('Button 컴포넌트 테스트', () => {
  it('버튼의 텍스트가 렌더링이 된다.', () => {
    const { getByText } = render(<Button>Click Me</Button>);
    expect(getByText('Click Me')).toBeInTheDocument();
  });

  it('버튼을 클릭하면 onClick 함수가 호출된다.', async () => {
    const user = userEvent.setup();
    const spy = vi.fn();
    const { getByText } = render(<Button onClick={spy}>onClick Event</Button>);

    await user.click(getByText('onClick Event'));
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('disabled 상태일 때 클릭이 동작하지 않는다', async () => {
    const user = userEvent.setup();
    const spy = vi.fn();
    const { getByText } = render(
      <Button onClick={spy} disabled>
        Disabled Button
      </Button>
    );

    await user.click(getByText('Disabled Button'));
    expect(spy).not.toHaveBeenCalled();
  });

  it('loading 상태일 때 스피너가 표시된다.', async () => {
    const { container } = render(<Button loading>Loading</Button>);

    expect(container.querySelector('.animate-spin')).toBeInTheDocument();
  });

  it('primary 스타일 적용된다.', () => {
    const variants = [
      'primary',
      'secondary',
      'outline',
      'gradient',
      'ghost',
      'text',
      'tabButton',
      'kakao',
      'popularButton',
      'nearbyButton',
      'google',
    ];

    variants.forEach((v) => {
      const { container } = render(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        <Button variant={v as any}>Variable Button</Button>
      );
      expect(container.firstChild).toMatchSnapshot(`button-${v}`);
    });
  });
});
