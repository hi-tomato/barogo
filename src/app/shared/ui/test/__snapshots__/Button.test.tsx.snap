// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-ghost 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed text-gray-600 hover:bg-gray-100 transition-colors px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-google 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-white hover:bg-gray-50 text-gray-700 border border-gray-300 hover:border-gray-400 transition-colors shadow-sm px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-gradient 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-gradient-to-r from-blue-500 to-blue-600 hover:shadow-md px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-kakao 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-[#FEE500] hover:bg-[#FDD835] text-black transition-colors px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-nearbyButton 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-[#1C4E80] hover:bg-[#154066] text-white rounded-full transition-colors px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-outline 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed border border-[#1C4E80]/15 text-[#1C4E80] hover:bg-[#1C4E80]/5 hover:border-[#1C4E80]/30 px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-popularButton 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-[#1C4E80] hover:bg-[#154066] text-white transition-colors px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-primary 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-[#2767a8] hover:bg-[#154066] text-white transition-colors px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-secondary 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-gray-500 hover:bg-gray-600 text-white transition-colors px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-tabButton 1`] = `
<button
  class="font-semibold transition-all duration-200 flex items-center justify-center space-x-2 disabled:opacity-50 disabled:cursor-not-allowed bg-[#1C4E80] text-white transition-colors px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;

exports[`Button 컴포넌트 테스트 > primary 스타일 적용된다. > button-text 1`] = `
<button
  class="font-semibold transition-all duration-200 disabled:opacity-50 disabled:cursor-not-allowed text-[#1C4E80] text-center transition-colors bg-transparent border-none px-4 py-2 text-base rounded-lg"
  type="button"
>
  <span>
    Variable Button
  </span>
</button>
`;
